src/core/execute.c

```diff
@@ -4943,6 +4996,8 @@ void exec_context_done(ExecContext *c) {
         c->apparmor_profile = mfree(c->apparmor_profile);
         c->smack_process_label = mfree(c->smack_process_label);
 
+        c->restrict_filesystems = set_free(c->restrict_filesystems);
```

Comment:
After removing the `set_new()` code as suggested by https://github.com/systemd/systemd/pull/18145#discussion_r690350651 and [changing bpf_lsm_parse_filesystems to initialize the set in there](https://github.com/systemd/systemd/pull/18145/files#diff-d37918508c319e9033e0f8281e72a1bd7f415d55377b42b3d3c7412c1a7b00c2R23) I'm seeing crashes here:

```gdb
Core was generated by `/sbin/init'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007ff52744101b in kill () from /usr/lib/libc.so.6
[Current thread is 1 (LWP 779)]
(gdb) bt
#0  0x00007ff52744101b in kill () from /usr/lib/libc.so.6
#1  0x00005563d444f854 in crash (sig=6) at ../src/core/main.c:262
#2  <signal handler called>
#3  0x00007ff527440d22 in raise () from /usr/lib/libc.so.6
#4  0x00007ff52742a862 in abort () from /usr/lib/libc.so.6
#5  0x00007ff527482d28 in __libc_message () from /usr/lib/libc.so.6
#6  0x00007ff52748a92a in malloc_printerr () from /usr/lib/libc.so.6
#7  0x00007ff52748b80c in unlink_chunk.constprop () from /usr/lib/libc.so.6
#8  0x00007ff52748c07b in _int_free () from /usr/lib/libc.so.6
#9  0x00007ff52748f9e8 in free () from /usr/lib/libc.so.6
#10 0x00007ff5278b7420 in _hashmap_clear (h=0x5563d5bd5f78, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:924
#11 0x00007ff5278b730d in _hashmap_free (h=0x5563d5bd5f78, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:888
#12 0x00005563d453b5cb in hashmap_free (h=0x5563d5bd5f78) at ../src/basic/hashmap.h:102
#13 0x00005563d453f006 in unit_clear_dependencies (u=0x5563d5c39c70) at ../src/core/unit.c:558
#14 0x00005563d453f7c0 in unit_free (u=0x5563d5c39c70) at ../src/core/unit.c:698
#15 0x00005563d44d2a78 in manager_clear_jobs_and_units (m=0x5563d5b04c30) at ../src/core/manager.c:1437
#16 0x00005563d44da840 in manager_reload (m=0x5563d5b04c30) at ../src/core/manager.c:3324
#17 0x00005563d4458ae8 in invoke_main_loop (m=0x5563d5b04c30, saved_rlimit_nofile=0x7fff74b947b0, saved_rlimit_memlock=0x7fff74b947c0, ret_reexecute=0x7fff74b946db,
    ret_retval=0x7fff74b946e0, ret_shutdown_verb=0x7fff74b94730, ret_fds=0x7fff74b94748, ret_switch_root_dir=0x7fff74b94720, ret_switch_root_init=0x7fff74b94728,
    ret_error_message=0x7fff74b94738) at ../src/core/main.c:1966
#18 0x00005563d445c5c8 in main (argc=1, argv=0x7fff74b949c8) at ../src/core/main.c:2942
```

I'm probably missing something obvious but I can't figure out why this happens.

Reply:
Sorry, pasted the wrong backtrace, this is the right one:

```
#0  0x00007fcb3c0a501b in kill () from /usr/lib/libc.so.6
#1  0x000055a50e8e9854 in crash (sig=6) at ../src/core/main.c:262
#2  <signal handler called>
#3  0x00007fcb3c0a4d22 in raise () from /usr/lib/libc.so.6
#4  0x00007fcb3c08e862 in abort () from /usr/lib/libc.so.6
#5  0x00007fcb3c0e6d28 in __libc_message () from /usr/lib/libc.so.6
#6  0x00007fcb3c0ee92a in malloc_printerr () from /usr/lib/libc.so.6
#7  0x00007fcb3c0f0408 in _int_free () from /usr/lib/libc.so.6
#8  0x00007fcb3c0f39e8 in free () from /usr/lib/libc.so.6
#9  0x00007fcb3c51965c in __unique_prefix_static_free_wrapper5 (a=0x55a50f18fb40) at ../src/basic/hash-funcs.c:13
#10 0x00007fcb3c51b3dd in _hashmap_clear (h=0x55a50f1f68a0, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:916
#11 0x00007fcb3c51b30d in _hashmap_free (h=0x55a50f1f68a0, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:888
#12 0x000055a50e918cb6 in set_free (s=0x55a50f1f68a0) at ../src/basic/set.h:20
#13 0x000055a50e9335ce in exec_context_done (c=0x55a50f192818) at ../src/core/execute.c:4999
#14 0x000055a50e9d94d3 in unit_done (u=0x55a50f1922e0) at ../src/core/unit.c:643
#15 0x000055a50e9d9595 in unit_free (u=0x55a50f1922e0) at ../src/core/unit.c:664
#16 0x000055a50e96ac40 in manager_dispatch_cleanup_queue (m=0x55a50f178c30) at ../src/core/manager.c:1170
#17 0x000055a50e9732f1 in manager_loop (m=0x55a50f178c30) at ../src/core/manager.c:3016
#18 0x000055a50e8f28d8 in invoke_main_loop (m=0x55a50f178c30, saved_rlimit_nofile=0x7ffcf08c9590, saved_rlimit_memlock=0x7ffcf08c95a0, ret_reexecute=0x7ffcf08c94bb,
    ret_retval=0x7ffcf08c94c0, ret_shutdown_verb=0x7ffcf08c9510, ret_fds=0x7ffcf08c9528, ret_switch_root_dir=0x7ffcf08c9500, ret_switch_root_init=0x7ffcf08c9508,
    ret_error_message=0x7ffcf08c9518) at ../src/core/main.c:1933
#19 0x000055a50e8f65c8 in main (argc=1, argv=0x7ffcf08c97a8) at ../src/core/main.c:2942
```

Reply:
The cause was https://github.com/systemd/systemd/pull/18145#discussion_r692757285